TESTING.md

```diff
@@ -0,0 +1,116 @@
+* Testing can prevent regressions by verifying that new changes don't break existing functionality.
+* Running tests locally accelerate the feedback loop, removing the dependency that contributors might have on CI when working on a Pull Request.
+
+This document will focus on teaching you about the different test suites that we currently have and how to run different scenarios to help your development experience!
+
+# Test categories
+
+## Unit tests
+
+Unit tests are used to test particular code snippets in isolation. They are your best ally when looking for quick feedback loops in a particular function.
+
+Imagine you're working on a PR that adds a new field to the ScrapeConfig CRD and you want to test if your change is reflected to the configmap. Instead of creating a full Kubernetes cluster, installing all the CRDs, running the Prometheus-Operator, deploying a Prometheus resource with ScrapeConfigSelectors and finally check if your change made it to the live object, you could simply write or extend a unit test for the configmap generation.
+
+Here is an example test that checks if the string generated from ScrapeConfigs are equal to an expected file.
+
+https://github.com/prometheus-operator/prometheus-operator/blob/cb534214415beb3c39353988ae85f2bb07f245e7/pkg/prometheus/promcfg_test.go#L4945-L4956
+
+Unit tests can be run with:
+
+```
+make test-unit
+```
+
+They can also be run for particular packages:
+
+```
+go test ./pkg/prometheus/server
+```
+
+Or even particular functions:
+
+```
+go test -run ^TestPodLabelsAnnotations$ ./pkg/prometheus/server
+```
+
+### Testing multiline string comparison - Golden files
+
+Golden files are plaintext documents designed to facilitate the validation of lengthy strings. They come in handy when, for instance, you need to test a Prometheus configuration that's generated using Go structures. You can marshal this configuration into YAML and then compare it against a static reference to ensure a match. Golden files offer an elegant solution to this challenge, sparing you the need to hard-code the static configuration directly into your test code.
+
+In the example below, we're generating the Prometheus configuration (which can easily have 100+ lines for each individual test) and comparing it against a golden file:
+
+https://github.com/prometheus-operator/prometheus-operator/blob/aeceb0b4fadc8307a44dc55afdceca0bea50bbb0/pkg/prometheus/promcfg_test.go#L102-L277
+
+If not for golden files, the test above, instead of ~150 lines, would easily require around ~1000 lines. The usage of golden files help us maintain test suites with several multiline strings comparison without sacrifing test readability.
+
+### Updating Golden Files
+
+There are contributions, e.g. adding a new required field to an existing configuration, that requires us to update several golden files at once. This can easily be done with the command below:
```

Comment:
(nit)
```suggestion
There are contributions, e.g. adding a new required field to an existing configuration, that require to update several golden files at once. This can easily be done with the command below:
```